File: src/main/java/com/thebluehats/server/game/enchants/MegaLongBow.java
package com.thebluehats.server.game.enchants;

import java.util.ArrayList;
import java.util.UUID;

import com.google.inject.Inject;
import com.thebluehats.server.game.managers.enchants.CustomEnchant;
import com.thebluehats.server.game.managers.enchants.CustomEnchantUtils;
import com.thebluehats.server.game.managers.enchants.EnchantGroup;
import com.thebluehats.server.game.managers.enchants.EnchantProperty;
import com.thebluehats.server.game.managers.enchants.Timer;
import com.thebluehats.server.game.utils.EnchantLoreParser;

import org.bukkit.Material;
import org.bukkit.entity.Arrow;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityShootBowEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

public class MegaLongBow implements CustomEnchant, Listener {
    private final EnchantProperty<Integer> amplifier = new EnchantProperty<>(1, 2, 3);

    private final CustomEnchantUtils customEnchantUtils;
    private final Timer<UUID> timer;

    @Inject
    public MegaLongBow(CustomEnchantUtils customEnchantUtils, Timer<UUID> timer) {
        this.customEnchantUtils = customEnchantUtils;
        this.timer = timer;
    }

    @EventHandler
    public void onArrowShoot(EntityShootBowEvent event) {
        if (event.getEntity() instanceof Player && event.getProjectile() instanceof Arrow) {
            Player player = (Player) event.getEntity();
            Arrow arrow = (Arrow) event.getProjectile();

            ItemStack bow = player.getInventory().getItemInHand();

            if (customEnchantUtils.itemHasEnchant(this, bow)) {
               execute(player, arrow, customEnchantUtils.getEnchantLevel(this, bow));
            }
        }
    }

    public void execute(Player player, Arrow arrow, int level) {
        UUID playerUuid = player.getUniqueId();

        if (!timer.isRunning(playerUuid)) {
            arrow.setCritical(true);
            arrow.setVelocity(player.getLocation().getDirection().multiply(2.90));
            player.addPotionEffect(new PotionEffect(PotionEffectType.JUMP, 40, amplifier.getValueAtLevel(level)), true);
        }

        timer.start(playerUuid, 20, false);
    }

    @Override
    public String getName() {
        return "Mega Longbow";
    }

    @Override
    public String getEnchantReferenceName() {
        return "Mlb";
    }

    @Override
    public ArrayList<String> getDescription(int level) {
        EnchantLoreParser enchantLoreParser = new EnchantLoreParser(
                "One shot per second, this bow is<br/>automatically fully drawn and<br/>grants <green>Jump Boost {0}</green> (2s)");

        enchantLoreParser.setSingleVariable("II", "III", "IV");

        return enchantLoreParser.parseForLevel(level);
    }

    @Override
    public boolean isDisabledOnPassiveWorld() {
        return false;
    }

    @Override
    public EnchantGroup getEnchantGroup() {
        return EnchantGroup.B;
    }

    @Override
    public boolean isRareEnchant() {
        return true;
    }

    @Override
    public Material[] getEnchantItemTypes() {
        return new Material[] { Material.BOW };
    }
}

File: src/main/kotlin/com/thebluehats/server/game/enchants/MegaLongBow.kt
package com.thebluehats.server.game.enchants

import com.google.inject.Inject
import com.thebluehats.server.game.managers.enchants.*
import com.thebluehats.server.game.managers.enchants.Timer
import com.thebluehats.server.game.utils.EnchantLoreParser
import org.bukkit.Material
import org.bukkit.entity.Arrow
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityShootBowEvent
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import java.util.*

class MegaLongBow @Inject constructor(
    private val customEnchantUtils: CustomEnchantUtils,
    private val timer: Timer<UUID>
) : CustomEnchant, Listener {
    private val amplifier = EnchantProperty(1, 2, 3)

    override val name: String get() = "Mega Longbow"
    override val enchantReferenceName: String get() = "mlb"
    override val isDisabledOnPassiveWorld: Boolean get() = false
    override val enchantGroup: EnchantGroup get() = EnchantGroup.B
    override val isRareEnchant: Boolean get() = true
    override val enchantItemTypes: Array<Material> get() = arrayOf(Material.BOW)

    @EventHandler
    fun onArrowShoot(event: EntityShootBowEvent) {
        if (event.entity is Player && event.projectile is Arrow) {
            val player = event.entity as Player
            val arrow = event.projectile as Arrow
            val bow = player.inventory.itemInHand

            val data = customEnchantUtils.getItemEnchantData(this, bow)

            if (data.itemHasEnchant()) execute(player, arrow, data.enchantLevel)
        }
    }

    fun execute(player: Player, arrow: Arrow, level: Int) {
        val playerUuid = player.uniqueId

        if (!timer.isRunning(playerUuid)) {
            arrow.isCritical = true
            arrow.velocity = player.location.direction.multiply(2.9)
            player.addPotionEffect(PotionEffect(PotionEffectType.JUMP, 40, amplifier.getValueAtLevel(level)), true)
        }

        timer.start(playerUuid, 20)
    }

    override fun getDescription(level: Int): ArrayList<String> {
        val enchantLoreParser = EnchantLoreParser(
            "One shot per second, this bow is<br/>automatically fully drawn and<br/>grants <green>Jump Boost {0}</green> (2s)"
        )

        enchantLoreParser.setSingleVariable("II", "III", "IV")

        return enchantLoreParser.parseForLevel(level)
    }
}